### PR TITLE
Refactor SSZ for Proper Caching, Thread-Safe Usage, and Efficiency

### LINKED ISSUE #31: Add Method to Find the Root of Serialized Data
Currently the only way to find the tree root of an object is by doing `HashTreeRoot` on the object. However if the object is already serialized there is no way to currently find the tree root of that object using the serialized data. The expected API would look something like this

```go
encodedObj,err := Marshal(object)
root,err := SerializedTreeRoot(encodedObj)

```

This issue will be resolved when this method is implemented

### LINKED ISSUE #36: SSZ Hash Cache Tests are Flakey
I currently see the following error when running tests with bazel in the repository:

```go
--- FAIL: TestBlockCache_maxSize (0.05s)
    hash_cache_test.go:127: Expected hash cache key size to be 10000, got 10025
FAIL
```
I suspect this has to do with some race conditions in the hash cache package leading to the problem. Tagging @shayzluf on this.

### LINKED ISSUE #49: Signing root panics for objects without a signature field
See https://github.com/prysmaticlabs/prysm/issues/2959 for an example.

### LINKED ISSUE #63: Change fmt.Errorf calls to errors.Wrap calls for easier tracing
This is a tracking issue for modifying all fmt.Errorf calls in the go-ssz repo to the `errors.Wrapf` pattern which offers more benefits in terms of tracing the error calls and their origin.

### PR DESCRIPTION
**Pull request by rauljordan**
This PR resolves #63, resolves #49, resolves #36, resolves #31.

## Background

SSZ in its current form is **not thread-safe**. This is a problem with its overall core design, which uses a **global cache** to store factories for marshalers/unmarshalers of different types. Calling hash tree root concurrently across routines would cause data races which lead to panics/deadlocks and are dangerous to have in production. Also, computing the unique key to store values in the global cache would be expensive, as we'd need to infer unique type and value information to do so at runtime.

## Changes

This PR removes the global cache and instead opts for a simple, recursive implementation of marshaling/unmarshaling/hashtreeroot for all different types. We introduce the `SSZAble` interface defined by:

```go
// SSZAble defines a type which can marshal/unmarshal and compute its
// hash tree root according to the Simple Serialize specification.
// See: https://github.com/ethereum/eth2.0-specs/blob/v0.8.2/specs/simple-serialize.md.
type SSZAble interface {
	Root(val reflect.Value, typ reflect.Type, maxCapacity uint64) ([32]byte, error)
	Marshal(val reflect.Value, typ reflect.Type, buf []byte, startOffset uint64) (uint64, error)
	Unmarshal(val reflect.Value, typ reflect.Type, buf []byte, startOffset uint64) (uint64, error)
}
```
and define structs for each major type that implement the interface, such as slices, structs, arrays, and more. Each struct that implements SSZAble has a local cache for its computed roots, which prevents race conditions from using a global store in the previous approach.

We delete the `ssz_utils_cache.go` file, which was the core factory that stored marshaler/unmarshaler functions in a map for O(1) access at runtime, as sacrificing the thread safety of the core SSZ design for speed is unacceptable. With these changes, we can be confident tests such as the following fully pass under the go race detector:

```go
func TestHashTreeRoot_ConcurrentAccess(t *testing.T) {
	item := &truncateSignatureCase{
		Slot:              10,
		PreviousBlockRoot: []byte{'a', 'b'},
		Signature:         []byte("TESTING23"),
	}
	var wg sync.WaitGroup
	// We ensure the hash tree root function can be computed in a thread-safe manner.
	wg.Add(100)
	for i := 0; i < 100; i++ {
		go func(tt *testing.T, w *sync.WaitGroup) {
			if _, err := HashTreeRoot(item); err != nil {
				tt.Fatal(err)
			}
			w.Done()
		}(t, &wg)
	}
	wg.Wait()
}
```

## Key Takeaways

Deadlock situations can be avoided if locking of shared resources is **only** done for non-recursive types. That is, using local caches for basic types, basic type slices, and basic type arrays will never lead to a deadlock situation, and these types account for the majority of SSZ's inefficiency anyways. 

### Master Branch

**Cache-enabled, `HashTreeRoot(MainnetState)`**
- First run: 500ms
- Subsequent runs: 200ms

### This Branch

**Cache-enabled, `HashTreeRoot(MainnetState)`**
- First run: 120ms
- Subsequent runs: 25ms

### CODE REVIEW

File: BUILD.bazel [hunk 1]
```diff
@@ -53,39 +53,20 @@ go_library(
...
-        "marshal_unmarshal_test.go",
-        "signing_root_test.go",
-        "struct_utils_test.go",
-        "marshal_test.go",
+        "round_trip_test.go",
+        "ssz_test.go",
     ],
     embed = [":go_default_library"],
-    deps = ["@com_github_minio_highwayhash//:go_default_library"],
+    race = "on",
```
Comment (rauljordan) [inline BUILD.bazel]: Make sure race detection is enabled

File: ssz.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
+package ssz
```
Comment (rauljordan) [inline ssz.go]: No need to review this file, simply collected `Marshal, Unmarshal, HashTreeRoot, SigningRoot` into a single file

File: ssz_test.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
+package ssz
```
Comment (rauljordan) [inline ssz_test.go]: No review needed here, simply consolidated the tests of `unmarshal_test.go, marshal_test.go, hash_tree_root_test.go, signing_root_test.go` into a single file

File: types/array_basic.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	hashCache map[string]interface{}
+}
+
+func newBasicArraySSZ() *basicArraySSZ {
+	return &basicArraySSZ{
+		hashCache: make(map[string]interface{}),
+	}
+}
+
+func (b *basicArraySSZ) Root(val reflect.Value, typ reflect.Type, maxCapacity uint64) ([32]byte, error) {
```
Comment (rauljordan) [inline types/array_basic.go]: Review

File: types/array_basic.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	if err != nil {
+		return [32]byte{}, err
+	}
+	if !bytes.Equal(hashKey, make([]byte, 32*numItems)) {
+		b.hashCache[string(hashKey)] = root
+	}
+	return root, nil
+}
+
+func (b *basicArraySSZ) Marshal(val reflect.Value, typ reflect.Type, buf []byte, startOffset uint64) (uint64, error) {
```
Comment (rauljordan) [inline types/array_basic.go]: Copied over from the previous marshal.go file, no need for review

File: types/array_basic.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	for i := 0; i < val.Len(); i++ {
+		index, err = factory.Marshal(val.Index(i), typ.Elem(), buf, index)
+		if err != nil {
+			return 0, err
+		}
+	}
+	return index, nil
+}
+
+func (b *basicArraySSZ) Unmarshal(val reflect.Value, typ reflect.Type, input []byte, startOffset uint64) (uint64, error) {
```
Comment (rauljordan) [inline types/array_basic.go]: Copied over from the previous unmarshal.go file, no need for review

File: types/array_composite.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	if err != nil {
+		return [32]byte{}, err
+	}
+	if val.Len() > 0 {
+		b.hashCache[string(hashKey)] = root
+	}
+	return root, nil
+}
+
+func (b *compositeArraySSZ) Marshal(val reflect.Value, typ reflect.Type, buf []byte, startOffset uint64) (uint64, error) {
```
Comment (rauljordan) [inline types/array_composite.go]: Copied over from the previous marshal.go file, no need for review

File: types/array_composite.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+		// We increase the offset indices accordingly.
+		currentOffsetIndex = nextOffsetIndex
+		fixedIndex += BytesPerLengthOffset
+	}
+	index = currentOffsetIndex
+	return index, nil
+}
+
+func (b *compositeArraySSZ) Unmarshal(val reflect.Value, typ reflect.Type, input []byte, startOffset uint64) (uint64, error) {
```
Comment (rauljordan) [inline types/array_composite.go]: Copied over from the previous unmarshal.go file, no need for review

File: types/factory.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
+package types
```
Comment (rauljordan) [inline types/factory.go]: New file, please review

File: types/helpers.go [hunk 10]
```diff
@@ -25,6 +25,18 @@ func init() {
 	}
 }
 
+// Given ordered BYTES_PER_CHUNK-byte chunks, if necessary utilize zero chunks so that the
+// number of chunks is a power of two, Merkleize the chunks, and return the root.
+// Note that merkleize on a single chunk is simply that chunk, i.e. the identity
+// when the number of chunks is one.
+func bitwiseMerkleize(chunks [][]byte, count uint64, limit uint64) ([32]byte, error) {
```
Comment (rauljordan) [inline types/helpers.go]: Using protolambda's implementation for speed here

File: types/slice_basic.go [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+	for i := 0; i < val.Len(); i++ {
+		index, err = factory.Marshal(val.Index(i), typ.Elem(), buf, index)
+		if err != nil {
+			return 0, err
+		}
+	}
+	return index, nil
+}
+
+func (b *basicSliceSSZ) Unmarshal(val reflect.Value, typ reflect.Type, input []byte, startOffset uint64) (uint64, error) {
```
Comment (rauljordan) [inline types/slice_basic.go]: Copied over from the previous unmarshal.go file, no need for review

File: types/slice_basic.go [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+	output := make([]byte, 32)
+	copy(output, buf.Bytes())
+	merkleRoot, err := bitwiseMerkleize(chunks, uint64(len(chunks)), limit)
+	if err != nil {
+		return [32]byte{}, err
+	}
+	return mixInLength(merkleRoot, output), nil
+}
+
+func (b *basicSliceSSZ) Marshal(val reflect.Value, typ reflect.Type, buf []byte, startOffset uint64) (uint64, error) {
```
Comment (rauljordan) [inline types/slice_basic.go]: Copied over from the previous marshal.go file, no need for review

File: types/slice_composite.go [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+	root, err := bitwiseMerkleize(chunks, uint64(len(chunks)), objLen)
+	if err != nil {
+		return [32]byte{}, err
+	}
+	result := mixInLength(root, output)
+	b.hashCache[string(hashKey)] = result
+	return result, nil
+}
+
+func (b *compositeSliceSSZ) Marshal(val reflect.Value, typ reflect.Type, buf []byte, startOffset uint64) (uint64, error) {
```
Comment (rauljordan) [inline types/slice_composite.go]: Copied over from the previous marshal.go file, no need for review

File: types/slice_composite.go [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+		// We increase the offset indices accordingly.
+		currentOffsetIndex = nextOffsetIndex
+		fixedIndex += BytesPerLengthOffset
+	}
+	index = currentOffsetIndex
+	return index, nil
+}
+
+func (b *compositeSliceSSZ) Unmarshal(val reflect.Value, typ reflect.Type, input []byte, startOffset uint64) (uint64, error) {
```
Comment (rauljordan) [inline types/slice_composite.go]: Copied over from the previous unmarshal.go file, no need for review

File: types/struct.go [hunk 15]
```diff
@@ -0,0 +1,321 @@
...
+		roots[i] = r[:]
+	}
+	root, err := bitwiseMerkleize(roots, uint64(numFields), uint64(numFields))
+	if err != nil {
+		return [32]byte{}, err
+	}
+	return root, nil
+}
+
+func (b *structSSZ) Marshal(val reflect.Value, typ reflect.Type, buf []byte, startOffset uint64) (uint64, error) {
```
Comment (rauljordan) [inline types/struct.go]: Copied over from the previous marshal.go file, no need for review

File: types/struct.go [hunk 16]
```diff
@@ -0,0 +1,321 @@
...
+
+			// We increase the offset indices accordingly.
+			currentOffsetIndex = nextOffsetIndex
+			fixedIndex += BytesPerLengthOffset
+		}
+	}
+	return currentOffsetIndex, nil
+}
+
+func (b *structSSZ) Unmarshal(val reflect.Value, typ reflect.Type, input []byte, startOffset uint64) (uint64, error) {
```
Comment (rauljordan) [inline types/struct.go]: Copied over from the previous unmarshal.go file, no need for review

File: deps.bzl [hunk 17]
```diff
@@ -106,6 +106,20 @@ def go_ssz_dependencies():
         importpath = "github.com/minio/highwayhash",
```
Comment (prestonvanloon) [inline deps.bzl]: Delete this, no longer used?

File: ssz.go [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+// HashTreeRootWithCapacity determines the root hash of a dynamic list
+// using SSZ's Merkleization and applies a max capacity value when computing the root.
+// If the input is not a slice, the function returns an error.
+//
+//  accountBalances := []uint64{1, 2, 3, 4}
+//  root, err := HashTreeRootWithCapacity(accountBalances, 100) // Max 100 accounts.
+//  if err != nil {
+//      return fmt.Errorf("failed to compute root: %v", err)
```
Comment (prestonvanloon) [inline ssz.go]:
```suggestion
//      return errors.Wrap(err, "failed to compute root")
```

File: ssz.go [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+//      Field2 []byte
+//  }
+//
+//  ex := exampleStruct{
+//      Field1: 10,
+//      Field2: []byte{1, 2, 3, 4},
+//  }
+//  root, err := HashTreeRoot(ex)
+//  if err != nil {
+//      return fmt.Errorf("failed to compute root: %v", err)
```
Comment (prestonvanloon) [inline ssz.go]:
```suggestion
//      return errors.Wrap(err, "failed to compute root")
```

File: ssz.go [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+	}
+	if rval.IsNil() {
+		return errors.New("cannot output to pointer of nil value")
+	}
+	factory, err := types.SSZFactory(rval.Elem(), rtyp.Elem())
+	if err != nil {
+		return err
+	}
+	if _, err := factory.Unmarshal(rval.Elem(), rval.Elem().Type(), input, 0); err != nil {
+		return fmt.Errorf("could not unmarshal input into type: %v, %v", rval.Elem().Type(), err)
```
Comment (prestonvanloon) [inline ssz.go]:
```suggestion
		return errors.Wrapf(err, "could not unmarshal input into type: %v", rval.Elem().Type())
```

File: ssz.go [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+		return nil, err
+	}
+	if rval.Type().Kind() == reflect.Ptr {
+		if _, err := factory.Marshal(rval.Elem(), rval.Elem().Type(), buf, 0 /* start offset */); err != nil {
+			return nil, fmt.Errorf("failed to marshal for type: %v", rval.Elem().Type())
+		}
+		return buf, nil
+	}
+	if _, err := factory.Marshal(rval, rval.Type(), buf, 0 /* start offset */); err != nil {
+		return nil, fmt.Errorf("failed to marshal for type: %v", rval.Type())
```
Comment (prestonvanloon) [inline ssz.go]: Include the error in the message?

File: ssz.go [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+				continue
+			}
+			totalFields++
+		}
+		return types.StructFactory.FieldsHasher(elem, elemType, totalFields-1)
+	}
+	totalFields := 0
+	for i := 0; i < valObj.Type().NumField(); i++ {
+		// We skip protobuf related metadata fields.
+		if strings.Contains(valObj.Type().Field(i).Name, "XXX") {
```
Comment (prestonvanloon) [inline ssz.go]: These protobuf fields start with `XXX_`, a field with `XXX` anywhere in the name is still valid. Can we change this to check if the field starts with `XXX_`?

File: types/array_basic.go [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+
+func newBasicArraySSZ() *basicArraySSZ {
+	return &basicArraySSZ{
+		hashCache: make(map[string]interface{}),
+	}
+}
+
+func (b *basicArraySSZ) Root(val reflect.Value, typ reflect.Type, maxCapacity uint64) ([32]byte, error) {
+	numItems := val.Len()
+	hashKey := make([]byte, 32*numItems)
```
Comment (prestonvanloon) [inline types/array_basic.go]: what is 32? define a constant?

**Pull request merged**